File: 0238-product-of-array-except-self/0238-product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int[] answer = new int[nums.length];
        int[] prefix_prod = new int[nums.length];
        int[] suffix_prod = new int[nums.length];
        
        prefix_prod[0] = 1;
        for(int i=1 ; i< nums.length ;i++){
            prefix_prod[i] = prefix_prod[i-1] * nums[i-1];
        }

        
        suffix_prod[nums.length-1] = 1;
        for(int i=nums.length-2; i >= 0  ;i--){
            suffix_prod[i] = suffix_prod[i+1] * nums[i+1];
        }

        for(int i=0 ; i< nums.length ; i++){
            answer[i] = prefix_prod[i] * suffix_prod[i];
        }
        return answer;
    }
}

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        prefix_prod = [0] * len(nums)
        suffix_prod = [0] * len(nums)
        answer = [0] * len(nums)

        prefix_prod[0] = 1
        for i in range(1,len(nums)):
            prefix_prod[i] = prefix_prod[i-1] * nums[i-1]

        
        suffix_prod[len(nums)-1] = 1
        for i in range(len(nums)-2,-1,-1):
            suffix_prod[i] = suffix_prod[i+1] * nums[i+1]

        for i in range(len(nums)):
            answer[i] = prefix_prod[i] * suffix_prod[i]

        return answer